### ISSUE TITLE
Find marker genes in sc workbench doesn't work following update

### ISSUE THREAD

**Issue by beamilon** [IGS/gEAR#318]
@jorvis 
I tried several scRNAseq dataset with primary analysis to Find marker genes and it returns an error.
It was ok before the update.
![image](https://user-images.githubusercontent.com/70719873/169409227-8cf3b382-721b-4cc9-a15f-12556bc51c2b.png)

**Comment by adkinsrs** [comment]
Relates to https://github.com/IGS/gEAR/issues/218

**Comment by beamilon** [comment]
@adkinsrs I don't think it relates to #218. After yesterday update, Find marker genes didn't work for any datasets. 218 was just about a specific dataset that needed to be replace to a new version.

**Comment by adkinsrs** [comment]
Oh.  it's a scanpy issue though.  I think I can change a parameter to an argument and it may work.

**Comment by beamilon** [comment]
It seems to be working now but the marker genes from primary analysis are different than before.

**Comment by adkinsrs** [comment]
I think it is working now because you went step-wise through the scRNA-seq workbench.  There is a line of code that says if the user selects primary analysis, look for their "unsaved" analysis, which I believe is why the results may be different.  Scanpy on the production server is on a more recent version (v1.8.2) so I believe updates in that package may have modified some of the code for the "rank_genes_groups" plotting function.

**Comment by beamilon** [comment]
I don't know why look for the unsaved analysis when selecting primary analysis. Was there a reason for that?

**Comment by adkinsrs** [comment]
I need to look more into the analsysis workbench code internals, but selecting "primary analysis" tells gEAR to look for the h5ad file in the default area. But if modifications are made, such as writing out analyses results, these results are written out to copied h5ad files in the special "user unsaved" area (where they can eventually be saved).

One issue is that the "user unsaved" files are in the temporary storage area, and this area is wiped whenever anyone restarts the Apache server. So I think @jorvis and I need to discuss this some.

**Comment by jorvis** [comment]
I've modified the apache config to no longer use private tmp, so it will now survive service restarts (but not SERVER restarts).

**Comment by adkinsrs** [comment]
With the Dying Tall and Short HCs, chicken (Heller) dataset, the reason why it is failing on the "Find Marker Genes" step during primary analysis is that there are about 100 GeneIDs that are duplicated across multiple gene symbols.  One example is GeneID 776816 which in this dataset is mapping to both ACTC2L and ACTG1.

The issue is that the scanpy function to get the top N genes per group maps the scores (logfoldchange, p-value) to the index of the gene dataframe, which in this case is the GeneID.  When scanpy tries to then plot the top 5 genes, for example, GeneID 776816 has the 3rd best score but maps to 2 genes, meaning that 6 genes are present for 5 scores, which causes the error.  You can actually get the "compare genes" step to work if you plot the top 1 gene.

So I guess the question is to either fix the dataset to deduplicate the gene ID to gene symbol mapping, or have find_marker_genes deduplicate, which will fix the issue but I cannot predict which gene symbol would be used for that GeneID that the scores are saved under.

**Comment by beamilon** [comment]
@jorvis 
@adkinsrs 
I just tried with 5dpf, zebrafish scRNA-seq, neuromast (Piotrowski) and Find marker gene doesn't work anymore
![image](https://user-images.githubusercontent.com/70719873/169655962-a3b10269-0524-49ac-aec5-f362b832c6fc.png)

**Comment by adkinsrs** [comment]
Fixed that just now.  I had left some testing code in the scanpy piece of code, but meant to remove all of it

**Comment by jorvis** [comment]
Related to #323 ?

**Comment by beamilon** [comment]
I don't think it is related. This issue is fixed. The one not working now is the visualization.

**Comment by adkinsrs** [comment]
@beamilon is this particular ticket still an issue?

**Comment by beamilon** [comment]
@adkinsrs It is working now
